server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java

```diff
@@ -1115,53 +1114,29 @@ private synchronized CommitSession finishPreparingMutations(long time) {
     return commitSession;
   }
 
-  public CommitSession prepareMutationsForCommit(TservConstraintEnv cenv, List<Mutation> mutations)
-      throws TConstraintViolationException {
-
-    ConstraintChecker cc = constraintChecker.derive();
-
-    List<Mutation> violators = null;
-    Violations violations = new Violations();
+  public TabletMutationPrepAttempt prepareMutationsForCommit(final TservConstraintEnv cenv,
+      final List<Mutation> mutations) {
     cenv.setExtent(extent);
-    for (Mutation mutation : mutations) {
-      Violations more = cc.check(cenv, mutation);
-      if (more != null) {
-        violations.add(more);
-        if (violators == null) {
-          violators = new ArrayList<>();
-        }
-        violators.add(mutation);
-      }
-    }
-
-    long time = tabletTime.setUpdateTimes(mutations);
-
-    if (!violations.isEmpty()) {
-
-      HashSet<Mutation> violatorsSet = new HashSet<>(violators);
-      ArrayList<Mutation> nonViolators = new ArrayList<>();
-
-      for (Mutation mutation : mutations) {
-        if (!violatorsSet.contains(mutation)) {
-          nonViolators.add(mutation);
-        }
-      }
-
-      CommitSession commitSession = null;
-
-      if (nonViolators.size() > 0) {
-        // if everything is a violation, then it is expected that
-        // code calling this will not log or commit
-        commitSession = finishPreparingMutations(time);
-        if (commitSession == null) {
-          return null;
-        }
+    final ConstraintChecker constraints = constraintChecker.derive();
+    final TabletMutationPrepAttempt attempt = new TabletMutationPrepAttempt();
```

Comment:
The existing code attempts to minimize the work done when no mutations violate a constraint.  It would be nice to carry this behavior forward, but I am not sure of the best way to do this.

Below is one possibility I thought of...

```java
Iterator<Mutation> miter = mutations.iterator(); //TODO reverse iterator would be more efficient when everything is removed
while(miter.hasNext()) {
  Violations violations = constraints.check(cenv, mutation);
  if(violations != null) {
      attempt.addViolator(mutation, violations);  // TODO this could allocate the internal list the first time called
     miter.remove();  // TODO is it ok to modify the list passed in??
  }
}

// if nothing violates, then the passed in list is not copied or modified...
attempt.setMutations(mutations);
```

Reply:
I checked each call to this function and there are cases where the original mutations list is still subsequently used, so we should avoid modifying it.

Reply:
> I checked each call to this function and there are cases where the original mutations list is still subsequently used, so we should avoid modifying it.

Ok, thanks for checking that.  I suppose that is why the existing code does something with a hashset when there are mutations.  Could possibly do something like the following.

```java
List<Mutation> vl = null;

for(Mutation mutation : mutations) {
  // TODO check for violations and if when one is found add to vl... creating vl if needed
}

if(vl.isEmpty()){
   // just use mutations list
} else {
   // copy mutations list and subtract violating mutations
}

```

Reply:
My only caution about this approach is that it makes it possible for the original mutation list to be modified if there are no violations and the list of non-violators is retrieved and modified. I don't think it's currently an issue, and if you're not concerned about it becoming an issue in the future, I'll make the change.

Reply:
> My only caution about this approach is that it makes it possible for the original mutation list to be modified if there are no violations 

Yeah that is a problem.  One way to address the issue is for the originator of the list to make it immutable.  Accumulo could benefit from using more immutable collections.